flang/lib/Optimizer/Transforms/RewriteLoop.cpp

```diff
@@ -44,8 +44,13 @@ class ScfLoopConv : public mlir::OpRewritePattern<fir::LoopOp> {
     auto step = loop.step();
     assert(low && high && step);
     // ForOp has different bounds semantics. Adjust upper bound.
-    auto diff = rewriter.create<mlir::SubIOp>(loc, high, low);
-    auto adjustUp = rewriter.create<mlir::AddIOp>(loc, diff, step);
+    auto one = rewriter.create<mlir::ConstantIndexOp>(loc, 1);
+    auto minusOne = rewriter.create<mlir::ConstantIndexOp>(loc, -1);
+    auto stepIsNegative =
+        rewriter.create<mlir::CmpIOp>(loc, mlir::CmpIPredicate::slt, step, one);
+    auto adjust =
+        rewriter.create<mlir::SelectOp>(loc, stepIsNegative, minusOne, one);
```

Comment:
It looks like there is no need to select between 1 and -1 because the step is required to be positive to create a sci::ForOp.

https://github.com/flang-compiler/f18-llvm-project/blob/fec3f2e41aa6004cb8655c018a8f6556ea2bf866/mlir/include/mlir/Dialect/SCF/SCFOps.td#L46-L52

Reply:
Thanks for bringing that up,  then there is a slight issue to handle somewhere:  Fortran DO loops may have negative steps and I believe that fir.loop also can have negative steps (@schweitzpgi ?) to support this.
This is a valid Fortran program:
```
  do i=10,1,-1
   print *, i
  end do
end
```

@schweitzpgi, that's a pretty annoying scf.for limitation regarding Fortran, it sounds like the loop bounds may have to be reversed during the rewrite in case of negative steps and the index inside the loop modified accordingly.

Looking into this more, the sfc.for constraint is actually enforced if the step is constant:
https://github.com/flang-compiler/f18-llvm-project/blob/fec3f2e41aa6004cb8655c018a8f6556ea2bf866/mlir/lib/Dialect/SCF/SCF.cpp#L71-L72

Reply:
Isn't loop reversal illegal if there are loop carried dependences?  Converting to a trip count loop should always be safe, although it must apparently count up rather than down, which is an additional annoyance.  Then hope that induction variable elimination eventually cleans things up.

Reply:
I propose reverting Zachary's first change for the moment.

Zachary is going to work on avoiding the use of scf.loop, which is the proper approach here.